### PR TITLE
[Doc] Update the docstring of distributed APIs.

### PR DESCRIPTION
**Pull request by zheng-da**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: docs/source/api/python/dgl.distributed.rst [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+---------------
+
+Initialization and Exit
+```````````````````````
+
+.. autosummary::
+    :toctree: ../../generated/
+
+    initialize
+    exit_client
```
Comment (classicsong) [inline docs/source/api/python/dgl.distributed.rst:18]: Do we need to expose exit_client to use?
It seems use do not need to care about this.

File: docs/source/api/python/dgl.distributed.rst [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+-----------------
+
+.. autoclass:: DistGraph
+
+.. autosummary::
+    :toctree: ../../generated/
+
+    node_split
+    edge_split
+
```
Comment (classicsong) [inline docs/source/api/python/dgl.distributed.rst:30]: find_edges
rank
number_of_nodes
number_of_edges

File: python/dgl/distributed/dist_graph.py [hunk 3]
```diff
@@ -394,11 +402,12 @@ def local_partition(self):
...
         DistGraph provides a global view of the distributed graph. Internally,
         it may contains a partition of the graph if it is co-located with
-        the server. If there is no co-location, this returns None.
+        the server. When servers and clients run on separate sets of machines,
+        this returns None.
 
         Returns
         -------
-        DGLHeterograph
+        DGLGraph
```
Comment (classicsong) [inline python/dgl/distributed/dist_graph.py:410]: DistGraph?

Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:438]: This returns a local graph.

File: python/dgl/distributed/dist_tensor.py [hunk 4]
```diff
@@ -24,36 +24,89 @@ def _default_init_data(shape, dtype):
...
+    if the first dimension matches the number of edges, DistTensor wll use the edge partition
+    policy. To determine the partition policy automatically, a DistGraph object has to be created.
+    Users can overwrite the rule by providing a partition policy directly.
+
+    A distributed tensor can be identified by a unique name or is anonymous.
+    When a distributed tensor has a name, the tensor can be persistent if persistent=True.
+    Normally, when a DistTensor object is destroyed, the distributed tensor in the system
+    is destroyed at the same time. However, a persistent tensor lives in the system even if
+    the DistTenor object is destroyed in the trainer process. DGL does not allow an anonymous
+    tensor to be persistent.
```
Comment (BarclayII) [inline python/dgl/distributed/dist_tensor.py:48]: A distributed tensor can be either named or anonymous.
    A named distributed tensor can be made persistent if ``persistent=True``

Also I don't quite understand when a persistent tensor gets destroyed from this part of docstring.

File: python/dgl/distributed/graph_services.py [hunk 5]
```diff
@@ -247,10 +247,9 @@ def _distributed_access(g, nodes, issue_remote_req, local_access):
 def sample_neighbors(g, nodes, fanout, edge_dir='in', prob=None, replace=False):
     """Sample from the neighbors of the given nodes from a distributed graph.
 
-    When sampling with replacement, the sampled subgraph could have parallel edges.
-
-    For sampling without replace, if fanout > the number of neighbors, all the
-    neighbors are sampled.
+    For each node, a number of inbound (or outbound when ``edge_dir == 'out'``) edges
+    will be randomly chosen.  The graph returned will then contain all the nodes in the
```
Comment (classicsong) [inline python/dgl/distributed/graph_services.py:251]: The graph returned will only contains the sampled edges.

File: python/dgl/distributed/graph_services.py [hunk 6]
```diff
@@ -260,27 +259,38 @@ def sample_neighbors(g, nodes, fanout, edge_dir='in', prob=None, replace=False):
...
+        neighboring edge of a node.  The feature must have only one element for each
+        edge.
+
+        The features must be non-negative floats, and the sum of the features of
+        inbound/outbound edges for every node must be positive (though they don't have
+        to sum up to one).  Otherwise, the result will be undefined.
     replace : bool, optional
         If True, sample with replacement.
 
+        When sampling with replacement, the sampled subgraph could have parallel edges.
```
Comment (classicsong) [inline python/dgl/distributed/graph_services.py:285]: parallel edges -> duplicated edges?

File: python/dgl/distributed/sparse_emb.py [hunk 7]
```diff
@@ -5,25 +5,52 @@
...
+    Because distributed embeddings are part of a model, they are updated by mini-batches.
+    The distributed embeddings have to be updated by DGL's optimizers instead of
+    the optimizers provided by the deep learning frameworks (e.g., Pytorch and MXNet).
+
+    To support efficient training on a graph with many nodes, the embeddings support sparse
+    updates. That is, only the embeddings involved in a mini-batch computation are updated.
+    Currently, DGL provides only one optimizer: `SparseAdagrad`. DGL will provide more
+    optimizers in the future.
+
+    Like distributed tensors, distributed embeddings are sharded and stored in a cluster
```
Comment (classicsong) [inline python/dgl/distributed/sparse_emb.py:21]: The distributed embeddings are sharded and stored in a cluster of machines and shared almost the same behavior as DistTensor, except it is trainable.

We do not need the following words.

File: python/dgl/distributed/sparse_emb.py [hunk 8]
```diff
@@ -5,25 +5,52 @@
...
-    >>> emb_init = lambda shape, dtype: F.zeros(shape, dtype, F.cpu())
-    >>> emb = dgl.distributed.DistEmbedding(g.number_of_nodes(), 10)
+    >>> def initializer(shape, dtype):
+            arr = th.zeros(shape, dtype=dtype)
+            arr.uniform_(-1, 1)
+            return arr
+    >>> emb = dgl.distributed.DistEmbedding(g.number_of_nodes(), 10, init_func=initializer)
     >>> optimizer = dgl.distributed.SparseAdagrad([emb], lr=0.001)
     >>> for blocks in dataloader:
     >>>     feats = emb(nids)
```
Comment (classicsong) [inline python/dgl/distributed/sparse_emb.py:56]: I think we also need to show optimizer.step() and etc.

File: python/dgl/distributed/sparse_emb.py [hunk 9]
```diff
@@ -88,15 +115,17 @@ def _init_state(shape, dtype):
...
 class SparseAdagrad:
-    ''' The Adagrad optimizer for sparse embeddings.
+    ''' The sparse Adagrad optimizer.
 
-    This optimizer collects gradients for the sparse embeddings and update
-    the embeddings in the distributed KVStore.
+    This optimizer implements a sparse version of the Adagrad algorithm.
+    It works with DistEmbedding and only update the embeddings
+    involved in a mini-batch to support efficient training on a graph with many
+    nodes and edges.
```
Comment (classicsong) [inline python/dgl/distributed/sparse_emb.py:123]: Add note here: grads are cleared after each step.

File: python/dgl/distributed/dist_context.py [hunk 10]
```diff
@@ -44,7 +44,23 @@ def _init_rpc(ip_config, num_servers, max_queue_size, net_type, role, num_thread
 def initialize(ip_config, num_servers=1, num_workers=0,
                max_queue_size=MAX_QUEUE_SIZE, net_type='socket',
                num_worker_threads=1):
-    """Init rpc service
+    """Initialize DGL's distributed module
+
+    This function has to be invoked at the very beginning of the training script.
```
[the remark this answers is not in the document]
Comment (aksnzhy) [reply python/dgl/distributed/dist_context.py:49]: It's better to say: Users cannot invoke any other distributed API before initializing.

File: python/dgl/distributed/dist_context.py [hunk 11]
```diff
@@ -138,7 +154,14 @@ def is_initialized():
...
 def exit_client():
-    """Register exit callback.
+    """Trainer exits
+
+    This function is called automatically when a Python process exits. Normally,
+    the training script does not need to invoke this function at the end.
+
+    In the case that the training script needs to initialize the distributed module
+    multiple times (so far, this is needed in the unit tests), the training script
+    needs to call `exit_client` before calling `initialize` again.
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_context.py:164]: Can we just hide this api?

File: python/dgl/distributed/dist_context.py [hunk 12]
```diff
@@ -44,7 +44,23 @@ def _init_rpc(ip_config, num_servers, max_queue_size, net_type, role, num_thread
 def initialize(ip_config, num_servers=1, num_workers=0,
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_context.py:44]: Can we use num_sampler_workers? Many users could be confused by num_worker and num_trainer.

File: python/dgl/distributed/dist_graph.py [hunk 13]
```diff
@@ -221,7 +221,7 @@ def __repr__(self):
 class DistGraphServer(KVServer):
     ''' The DistGraph server.
 
-    This DistGraph server loads the graph data and sets up a service so that clients can read data
+    This DistGraph server loads the graph data and sets up a service so that trainers can read data
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_graph.py:224]: trainers and samplers can read..

File: python/dgl/distributed/dist_context.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/distributed/dist_context.py:49]: The term `distributed API` is still not accurate enough because it includes both DGL distributed API and *framework-specific* distributed API.  You can be more specific on this term (something like "You need to initialize the module before calling any DGL distributed API or framework-specific distributed API.").

Also would you care explaining the reason why this has to be called before `init_process_group`?  And how about MXNet or Tensorflow (it's OK if you don't know)?

File: python/dgl/distributed/dist_context.py [hunk 14]
```diff
@@ -44,7 +44,23 @@ def _init_rpc(ip_config, num_servers, max_queue_size, net_type, role, num_thread
...
+    For example, when used with Pytorch, this function has to be invoked before
+    Pytorch's `pytorch.distributed.init_process_group`.
+
+    The initialization function acts differently in different modes. When the training
+    scripts runs as a trainer process, this API builds connections with DGL servers and
+    creates sampler processes; when the training script runs as a server process,
+    this API runs the server code and never returns.
+
+    DGL uses multiprocessing to parallelize distributed sampling. The sampling processes
+    have to be created in advance. `num_workers` specifies the number of sampling worker
```
Comment (BarclayII) [inline python/dgl/distributed/dist_context.py:59]:
```suggestion
    DGL uses multiprocessing to parallelize distributed sampling. You must create the sampling
    processes in advance. `num_workers` specifies the number of sampling worker
```

Also what does `in advance` precisely mean?  Maybe providing a link to a user guide/tutorial/example is helpful.

**Comment by BarclayII** [comment]
Also do you plan to provide the instructions for how users can create their own partition files for training with DGL without calling `dgl.distributed.partition_graph`?  This is relevant for users who wish to customize partitioning strategies or users who cannot run `partition_graph` since the graph cannot fit into host memory.

File: python/dgl/distributed/dist_context.py (hunk 12 above)
Comment (zheng-da) [reply python/dgl/distributed/dist_context.py:44]: we could say `num_samplers`. I'm thinking if the worker processes will only be used by samplers in the future?

File: python/dgl/distributed/dist_context.py [hunk 15]
```diff
@@ -44,7 +44,23 @@ def _init_rpc(ip_config, num_servers, max_queue_size, net_type, role, num_thread
...
+    processes per trainer process.
+
+    Users also have to provide the number of server processes on each machine in order
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/dist_context.py:62]: why not use "users have to"?

File: python/dgl/distributed/dist_context.py (hunk 11 above)
Comment (zheng-da) [reply python/dgl/distributed/dist_context.py:166]: this is docstring. it's not listed in the rst file.

File: python/dgl/distributed/dist_graph.py [hunk 16]
```diff
@@ -297,35 +297,72 @@ def start(self):
...
+    '''The class for accessing a distributed graph.
 
-    This provides the graph interface to access the partitioned graph data for distributed GNN
-    training. All data of partitions are loaded by the DistGraph server.
+    This class provides a subset of DGLGraph APIs for accessing partitioned graph data in
+    distributed GNN training and inference. Thus, its main use case is to work with
+    distributed sampling APIs to generate mini-batches and perform forward and
+    backward computation on the mini-batches.
+
+    **Note**: DistGraph currently only supports graphs with only one node type and one edge type.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:307]: I don't want to use "homogeneous graph". Technically, the API also works for heterogeneous graphs. Users need to keep the node type and edge type as node data and edge data.

File: python/dgl/distributed/dist_graph.py [hunk 17]
```diff
@@ -297,35 +297,72 @@ def start(self):
...
 
     DistGraph can run in two modes: the standalone mode and the distributed mode.
 
     * When a user runs the training script normally, DistGraph will be in the standalone mode.
-    In this mode, the input graph has to be constructed with only one partition. This mode is
-    used for testing and debugging purpose.
+      In this mode, the input data is constructed by
+      :py:meth:`~dgl.distributed.partition.partition_graph` with only one partition. This mode is
+      used for testing and debugging purpose. In this mode, users have to provide `part_config` so
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:314]: why double back quote? if we mention a symbol in the docstring, should we use back quote to mark it as code? for example, the line below mentions DistGraph.

File: python/dgl/distributed/dist_graph.py [hunk 18]
```diff
@@ -297,35 +297,72 @@ def start(self):
...
-    When running in the distributed mode, `DistGraph` uses shared-memory to access
-    the partition data in the local machine.
-    This gives the best performance for distributed training when we run `DistGraphServer`
-    and `DistGraph` on the same machine. However, a user may want to run them in separate
-    machines. In this case, a user may want to disable shared memory by passing
+    Users may want to run DistGraph servers and clients on separate sets of machines.
+    In this case, a user may want to disable shared memory by passing
     `disable_shared_mem=False` when creating `DistGraphServer`. When shared-memory is disabled,
-    a user has to pass a partition book.
+    a user has to pass a partition book. This is currently used for the testing purpose.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:328]: this can be useful if people want to run servers and clients on separate machines. however, this functionality hasn't been fully tested.

File: python/dgl/distributed/dist_graph.py [hunk 19]
```diff
@@ -297,35 +297,72 @@ def start(self):
...
-    part_config : str
-        The partition config file. It's used in the standalone mode.
+        The name of the graph. This name has to be the same as the one used for
+        partitioning a graph in :py:meth:`dgl.distributed.partition.partition_graph`.
+    gpb : GraphPartitionBook, optional
+        The partition book object. Normally, users do not need to provide the partition book.
+        This is mainly used for the testing purpose.
+    part_config : str, optional
+        The path of partition configuration file generated by
+        :py:meth:`dgl.distributed.partition.partition_graph`. It's used in the standalone mode.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:340]: we don't need to. we can get this information from the servers.

File: python/dgl/distributed/dist_graph.py [hunk 20]
```diff
@@ -499,34 +541,101 @@ def etypes(self):
         return ['_E']
 
     def number_of_nodes(self):
-        """Return the number of nodes"""
+        """Return the total number of nodes in the distributed graph.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:544]: i want to make it consistent with `number_of_nodes` in DGLGraph.

File: python/dgl/distributed/dist_graph.py [hunk 21]
```diff
@@ -499,34 +541,101 @@ def etypes(self):
...
+        --------
+        >>> g = dgl.distributed.DistGraph('ogb-product')
+        >>> print(g.number_of_nodes())
+        123718280
+        """
         return self._num_edges
 
     def node_attr_schemes(self):
-        """Return the node feature and embedding schemes."""
+        """Return the node feature schemes.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:576]: we can. why?

File: python/dgl/distributed/graph_partition_book.py [hunk 22]
```diff
@@ -77,10 +77,183 @@ def get_shared_mem_partition_book(graph_name, graph_part):
...
+            partition IDs
+        """
+
+    def partid2nids(self, partid):
+        """From partition id to global node IDs
+
+        Parameters
+        ----------
+        partid : int
+            partition id
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/graph_partition_book.py:186]: no, it's partition ID

File: python/dgl/distributed/graph_services.py (hunk 6 above)
Comment (zheng-da) [reply python/dgl/distributed/graph_services.py:285]: i copied from sample_neighbors. we should use a consistent term for this kind of edges. @BarclayII

File: python/dgl/distributed/dist_graph.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply python/dgl/distributed/dist_graph.py:548]: > i want to make it consistent with `number_of_nodes` in DGLGraph.

There is num_nodes and num_edges properties in DGLGraph right now.

File: python/dgl/distributed/dist_tensor.py [hunk 23]
```diff
@@ -24,36 +24,89 @@ def _default_init_data(shape, dtype):
...
+    or edges. When a distributed tensor is created, the partition policy is automatically
+    determined based on the first dimension if the partition policy is not provided: if
+    the first dimension matches the number of nodes, DistTensor will use the node partition policy;
+    if the first dimension matches the number of edges, DistTensor wll use the edge partition
+    policy. To determine the partition policy automatically, a DistGraph object has to be created.
+    Users can overwrite the rule by providing a partition policy directly.
+
+    A distributed tensor can be identified by a unique name or is anonymous.
+    When a distributed tensor has a name, the tensor can be persistent if persistent=True.
+    Normally, when a DistTensor object is destroyed, the distributed tensor in the system
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/dist_tensor.py:45]: yes. it'll free memory on the server. otherwise, there will be memory leak.

**Comment by zheng-da** [comment]
@BarclayII for your last question, it should appear in the user guide.
